File: src/main/java/com/vc/medium/L666.scala
package com.vc.medium

object L666 {
  def pathSum(arr: Array[Int]): Int = {
    import scala.collection.mutable
    val tree = new mutable.HashMap[Int, Int]()
    var total = 0
    def traverse(node: Int, preSum: Int): Unit = {
      val level = node / 10
      val pos = node % 10

      val currentSum = preSum + tree(node)
      val left = (level + 1) * 10 + 2 * pos - 1
      val right = (level + 1) * 10 + 2 * pos

      if(!tree.contains(left) && !tree.contains(right)) total += currentSum

      if(tree.contains(left)) traverse(left, currentSum)
      if(tree.contains(right)) traverse(right, currentSum)
    }
    arr.foreach(x => {
      val pos = x / 10
      val value = x % 10
      tree.put(pos, value)
    })
    traverse(arr(0) / 10, 0)
    total
  }
}

File: src/main/java/com/vc/medium/L666.java
package com.vc.medium;

import java.util.HashMap;

class L666 {
    private HashMap<Integer, Integer> tree;
    private int res = 0;
    public int pathSum(int[] nums) {
        this.tree = new HashMap<>();
        for(int i = 0; i < nums.length; i++) {
            int node = nums[i] / 10;
            int value = nums[i] % 10;
            tree.put(node, value);
        }
        helper(nums[0] / 10, 0);
        return res;
    }

    private void helper(int node, int sum) {
        if(!tree.containsKey(node)) return;

        int depth = node / 10;
        int pos = node % 10;

        int left = (depth + 1) * 10 + (2 * pos - 1);
        int right = left + 1;

        int total = tree.get(node) + sum;

        if(!tree.containsKey(left) && !tree.containsKey(right)) res += total;
        else {
            helper(left, total);
            helper(right, total);
        }
    }
}
